File: sdk-js/src/gatewayClient.ts
/**
 * Caller-side gateway polling for async jobs.
 */

import { DEFAULT_API_URL } from './constants';
import { resolveGatewayPathPrefix } from './pathPrefixes';
import { resolveBaseUrl } from './urls';

function normalizePrefix(prefix: string): string {
  if (!prefix) return '';
  const p = prefix.startsWith('/') ? prefix : `/${prefix}`;
  return p.replace(/\/$/, '');
}

function buildUrl(origin: string, gatewayPathPrefix: string, suffix: string): string {
  const base = resolveBaseUrl(origin, DEFAULT_API_URL);
  return `${base}${normalizePrefix(gatewayPathPrefix)}${suffix}`;
}

export interface GatewayPollResult {
  ok: boolean;
  status: number;
  body: string;
}

/**
 * GET .../requests/{requestId}/status with Bearer service key.
 */
export async function getRequestStatus(params: {
  /** API origin; defaults to `https://api.tollara.ai`. */
  baseUrl?: string | null;
  requestId: string;
  serviceKey: string;
  fetch?: typeof globalThis.fetch;
}): Promise<GatewayPollResult> {
  const { baseUrl, requestId, serviceKey, fetch: fetchFn = fetch } = params;
  const url = buildUrl(
    baseUrl ?? DEFAULT_API_URL,
    resolveGatewayPathPrefix(baseUrl),
    `/requests/${requestId}/status`,
  );
  try {
    const res = await fetchFn(url, {
      method: 'GET',
      headers: { Authorization: `Bearer ${serviceKey}` },
    });
    const body = await res.text();
    return { ok: res.ok, status: res.status, body };
  } catch {
    return { ok: false, status: 0, body: '' };
  }
}

/**
 * GET .../requests/{requestId}/result with Bearer service key.
 */
export async function getRequestResult(params: {
  /** API origin; defaults to `https://api.tollara.ai`. */
  baseUrl?: string | null;
  requestId: string;
  serviceKey: string;
  fetch?: typeof globalThis.fetch;
}): Promise<GatewayPollResult> {
  const { baseUrl, requestId, serviceKey, fetch: fetchFn = fetch } = params;
  const url = buildUrl(
    baseUrl ?? DEFAULT_API_URL,
    resolveGatewayPathPrefix(baseUrl),
    `/requests/${requestId}/result`,
  );
  try {
    const res = await fetchFn(url, {
      method: 'GET',
      headers: { Authorization: `Bearer ${serviceKey}` },
    });
    const body = await res.text();
    return { ok: res.ok, status: res.status, body };
  } catch {
    return { ok: false, status: 0, body: '' };
  }
}

File: sdk-dotnet/GatewayClient.cs
using System.Net.Http;
using System.Net.Http.Headers;

namespace AgentVend;

/// <summary>Caller-side gateway polling (sdk-api-spec §1.3–1.4).</summary>
public static class GatewayClient
{
    /// <summary>Poll status using the SDK default API origin and gateway path prefix (same as <see cref="AgentVendClient"/>).</summary>
    public static Task<(bool Ok, int StatusCode, string Body)> GetRequestStatusAsync(HttpClient http, string requestId,
        string agentKey, CancellationToken ct = default) =>
        GetRequestStatusAsync(http, AgentVendClient.DefaultApiUrl, AgentVendClient.DefaultGatewayPathPrefix, requestId,
            agentKey, ct);

    /// <summary>Poll status against explicit gateway base and path prefix (for custom or local stacks).</summary>
    public static async Task<(bool Ok, int StatusCode, string Body)> GetRequestStatusAsync(
        HttpClient http,
        string gatewayBaseUrl,
        string gatewayPathPrefix,
        string requestId,
        string agentKey,
        CancellationToken ct = default)
    {
        var url = BuildUrl(gatewayBaseUrl, gatewayPathPrefix, $"/requests/{requestId}/status");
        return await GetAsync(http, url, agentKey, ct);
    }

    /// <summary>Fetch result using the SDK default API origin and gateway path prefix (same as <see cref="AgentVendClient"/>).</summary>
    public static Task<(bool Ok, int StatusCode, string Body)> GetRequestResultAsync(HttpClient http, string requestId,
        string agentKey, CancellationToken ct = default) =>
        GetRequestResultAsync(http, AgentVendClient.DefaultApiUrl, AgentVendClient.DefaultGatewayPathPrefix, requestId,
            agentKey, ct);

    /// <summary>Fetch result against explicit gateway base and path prefix (for custom or local stacks).</summary>
    public static async Task<(bool Ok, int StatusCode, string Body)> GetRequestResultAsync(
        HttpClient http,
        string gatewayBaseUrl,
        string gatewayPathPrefix,
        string requestId,
        string agentKey,
        CancellationToken ct = default)
    {
        var url = BuildUrl(gatewayBaseUrl, gatewayPathPrefix, $"/requests/{requestId}/result");
        return await GetAsync(http, url, agentKey, ct);
    }

    private static async Task<(bool Ok, int StatusCode, string Body)> GetAsync(
        HttpClient http, string url, string agentKey, CancellationToken ct)
    {
        using var req = new HttpRequestMessage(HttpMethod.Get, url);
        req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", agentKey);
        var res = await http.SendAsync(req, ct);
        var body = await res.Content.ReadAsStringAsync(ct);
        return (res.IsSuccessStatusCode, (int)res.StatusCode, body);
    }

    private static string BuildUrl(string gatewayBaseUrl, string gatewayPathPrefix, string suffix)
    {
        var baseUrl = (gatewayBaseUrl ?? "").TrimEnd('/');
        var prefix = NormalizePrefix(gatewayPathPrefix);
        return baseUrl + prefix + suffix;
    }

    private static string NormalizePrefix(string? gatewayPathPrefix)
    {
        if (string.IsNullOrEmpty(gatewayPathPrefix)) return "";
        var p = gatewayPathPrefix.StartsWith('/') ? gatewayPathPrefix : "/" + gatewayPathPrefix;
        return p.TrimEnd('/');
    }
}
